pkg/controller/networkpolicy/controller_querier_replier.go

```diff
@@ -0,0 +1,182 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package networkpolicy provides NetworkPolicyController implementation to manage
+// and synchronize the Pods and Namespaces affected by Network Policies and enforce
+// their rules.
+
+package networkpolicy
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	networkingv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/networking/v1beta1"
+	"github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy/store"
+	antreatypes "github.com/vmware-tanzu/antrea/pkg/controller/types"
+)
+
+type EndpointQuerier interface {
+	QueryNetworkPolicies(namespace string, podName string) (*EndpointQueryResponse, error)
+}
+
+// EndpointQueryReplier is responsible for handling query requests from antctl query
+type EndpointQueryReplier struct {
+	networkPolicyController *NetworkPolicyController
+}
+
+// EndpointQueryResponse is the reply struct for QueryNetworkPolicies
+type EndpointQueryResponse struct {
+	Endpoints []Endpoint `json:"endpoints,omitempty"`
+}
+
+// Endpoint holds response information for an endpoint following a query
+type Endpoint struct {
+	Namespace string   `json:"namespace,omitempty"`
+	Name      string   `json:"name,omitempty"`
+	Policies  []Policy `json:"policies,omitempty"`
+	Rules     []Rule   `json:"rules,omitempty"`
+}
+
+type PolicyRef struct {
+	Namespace string    `json:"namespace,omitempty"`
+	Name      string    `json:"name,omitempty"`
+	UID       types.UID `json:"uid,omitempty"`
+}
+
+// Policy holds network policy information to be relayed to client following query endpoint
+type Policy struct {
+	PolicyRef
+	selector metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+type Rule struct {
+	PolicyRef
+	Direction networkingv1beta1.Direction `json:"direction,omitempty"`
+	RuleIndex int                         `json:"ruleindex,omitempty"`
+}
+
+// NewNetworkPolicyController returns a new *NetworkPolicyController.
+func NewEndpointQueryReplier(networkPolicyController *NetworkPolicyController) *EndpointQueryReplier {
+	n := &EndpointQueryReplier{
+		networkPolicyController: networkPolicyController,
+	}
+	return n
+}
+
+//QueryNetworkPolicies returns kubernetes network policy references relevant to the selected network endpoint. Relevant
+//policies fall into three categories: applied policies (Policies in Endpoint type) are policies which directly apply to
+//an endpoint, egress and ingress rules (Rules in Endpoint type) are policies which reference the endpoint in an ingress/
+//egress rule respectively.
+func (eq EndpointQueryReplier) QueryNetworkPolicies(namespace string, podName string) (*EndpointQueryResponse, error) {
+	// check if namespace and podName select an existing pod
+	_, err := eq.networkPolicyController.podInformer.Lister().Pods(namespace).Get(podName)
+	if err != nil {
+		return nil, nil
```

Comment:
do you want to return the error in this case?

Reply:
A nil error with a nil return value is handled as a specific case in the handler. It was a while ago but I believe we had a conversation about this @antoninbas. Should I handle the specific error rather than return nil?

Reply:
I think the current code is fine. Maybe the function comment should explicitly call out that we return `nil` and no error if the Namespace and Pod name do not match a known Pod.